code/contracts/ONEWallet.sol

```diff
@@ -351,25 +393,40 @@ contract ONEWallet is TokenTracker {
         return uint32(block.timestamp) - commitTime < REVEAL_MAX_DELAY;
     }
 
-    function _verifyReveal(bytes32 hash, uint32 indexWithNonce, bytes32 paramsHash) view internal
+    /// This function verifies that the first valid entry with respect to the given `eotp` in `commitLocker[hash]` matches the provided `paramsHash` and `verificationHash`. An entry is valid with respect to `eotp` iff `h3(entry.paramsHash . eotp)` equals `entry.verificationHash`
+    function _verifyReveal(bytes32 hash, uint32 indexWithNonce, bytes32 paramsHash, bytes32 eotp) view internal returns (uint32)
     {
-        Commit storage c = commitLocker[hash];
-        require(c.timestamp > 0, "Cannot find commit");
         uint32 index = indexWithNonce / maxOperationsPerInterval;
         uint8 nonce = uint8(indexWithNonce % maxOperationsPerInterval);
-        uint32 counter = c.timestamp / interval - t0;
-        require(counter == index, "Index - timestamp mismatch");
-        uint8 expectedNonce = nonces[counter];
-        require(nonce >= expectedNonce, "Nonce too low");
-        require(!c.completed, "Commit already completed");
-        require(c.paramsHash == paramsHash, "Invalid params hash");
-        // this should not happen (since old commit should be cleaned up already)
-        require(_isRevealTimely(c.timestamp), "Reveal too late");
+        Commit[] storage cc = commitLocker[hash];
+        require(cc.length > 0, "No commit found");
+        for (uint32 i = 0; i < cc.length; i++) {
+            Commit storage c = cc[i];
+            bytes32 expectedVerificationHash = keccak256(bytes.concat(c.paramsHash, eotp));
+            if (c.verificationHash != expectedVerificationHash) {
+                // Invalid entry. Ignore
+                continue;
+            }
+            require(c.paramsHash == paramsHash, "Parameter hash mismatch");
+            uint32 counter = c.timestamp / interval - t0;
+            require(counter == index, "Index - timestamp mismatch");
+            uint8 expectedNonce = nonces[counter];
+            require(nonce >= expectedNonce, "Nonce too low");
```

Comment:
Why do you enforce this condition? It might have negative consequences on the use case of the shared wallet that you mentioned before. I'd say that since each nonce corresponds to its own EOTP there is no need for such a constraint. EOTPs aggregated by the Merkle enable selective acknowledgment (see SmartOTPs paper.)

Reply:
Yes, this is a remnant of an early version to prevent some types of attacks. It should be modified to `require(nonce < maxOperationsPerInterval, "Nonce is invalid");`

Reply:
My previous comment in this thread is wrong. It would always be true that `nonce < maxOperationsPerInterval ` since nonce is the remainder of modulus `maxOperationsPerInterval`.  I think the only remaining purpose now for nonce is to track how many operations are already performed in the current time interval. The difference this line makes is whether we enforce nonce to be monotonically increasing. Suppose if we remove this line, then clients (possessing proofs) would be free to choose which nonce to use. It would be chaotic, since clients would have no idea what nonces other clients are using. If two clients choose the same nonce at the same time, one of them would fail. So in that situation, if two clients are trying nonces incrementally, it would not really be any better than the current situation enforced by this line: two clients would be forced to query the current nonce of the current block, and one of them would be forced to wait until the other finishes. On the other hand, if two clients coordinate with each other and try different nonces at the same time, then both transactions could go through. But then the client implementation would get complicated - WebSocket or WebRTC would have to be used for coordination, and we would inevitably need a server to broker the communication before the clients connect with each other. In view of that, I don't think we should remove the current constraint imposed by this line, until we do have proper WebSocket or WebRTC implementations to coordinate multiple clients. For now, the user(s) should coordinate manually and let one transaction finishes before the other in the multi-client use case. This approach is better for security reason as well - an attacker with the capability to intercept OTPs and read client side data would be forced to block the commit from the user's client, as otherwise the user's transaction would race with the attacker's. It makers this kind of attack, which is already very difficult to pull off, even easier to detect and harder to succeed.